File: frontend/lib/types/revenue.ts
export type RevenueHeadline = {
  window_days: number;
  computed_at: string;
  mrr_usd: number;
  arr_usd: number;
  arpu_usd: number;
  free_to_paid_conversion: number;
  revenue_per_user_usd: number;
  ltv_proxy_usd: number;
  churn_rate: number;
  paying_user_retention_d30: number;
};

export type RevenueFunnelStep = {
  key: string;
  label: string;
  users: number;
  conversion_from_prev: number;
  dropoff_from_prev: number;
};

export type RevenueFunnel = {
  steps: RevenueFunnelStep[];
};

export type RevenueSource = {
  source: string;
  acquired_users: number;
  paid_users: number;
  conversion_rate: number;
  mrr_usd: number;
  arr_usd: number;
};

export type RevenueCohort = {
  cohort_week_start: string;
  source: string;
  plan_tier: string;
  users: number;
  paid_users: number;
  revenue_usd: number;
  retention_d30: number | null;
  conversion_lag_days: number | null;
};

export type RevenueExperimentVariant = {
  experiment_key: string;
  variant: string;
  views: number;
  interactions: number;
  upgrades: number;
  paid_users: number;
  conversion_rate: number;
  revenue_per_user_usd: number;
  retention_d30: number;
};

export type RevenueChurnSignal = {
  churn_risk_users: number;
  canceled_users: number;
  inactive_paying_users: number;
  generated_at: string;
};

export type RevenueDashboard = {
  headline: RevenueHeadline;
  funnel: RevenueFunnel;
  funnel_by_source: RevenueSource[];
  revenue_by_source: RevenueSource[];
  paywall_performance: RevenueExperimentVariant[];
  cohorts: RevenueCohort[];
  churn_signals: RevenueChurnSignal;
};

File: backend/app/modules/analytics/model/revenue.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field


class RevenueHeadlineRead(BaseModel):
    window_days: int
    computed_at: datetime
    mrr_usd: float
    arr_usd: float
    arpu_usd: float
    free_to_paid_conversion: float
    revenue_per_user_usd: float
    ltv_proxy_usd: float
    churn_rate: float
    paying_user_retention_d30: float


class RevenueFunnelStepRead(BaseModel):
    key: str
    label: str
    users: int
    conversion_from_prev: float
    dropoff_from_prev: float


class RevenueFunnelRead(BaseModel):
    steps: list[RevenueFunnelStepRead] = Field(default_factory=list)


class RevenueSourceRead(BaseModel):
    source: str
    acquired_users: int
    paid_users: int
    conversion_rate: float
    mrr_usd: float
    arr_usd: float


class RevenueCohortRead(BaseModel):
    cohort_week_start: date
    source: str
    plan_tier: str
    users: int
    paid_users: int
    revenue_usd: float
    retention_d30: float | None = None
    conversion_lag_days: float | None = None


class RevenueExperimentVariantRead(BaseModel):
    experiment_key: str
    variant: str
    views: int
    interactions: int
    upgrades: int
    paid_users: int
    conversion_rate: float
    revenue_per_user_usd: float
    retention_d30: float


class RevenueChurnSignalRead(BaseModel):
    churn_risk_users: int
    canceled_users: int
    inactive_paying_users: int
    generated_at: datetime


class RevenueDashboardRead(BaseModel):
    headline: RevenueHeadlineRead
    funnel: RevenueFunnelRead
    funnel_by_source: list[RevenueSourceRead] = Field(default_factory=list)
    revenue_by_source: list[RevenueSourceRead] = Field(default_factory=list)
    paywall_performance: list[RevenueExperimentVariantRead] = Field(default_factory=list)
    cohorts: list[RevenueCohortRead] = Field(default_factory=list)
    churn_signals: RevenueChurnSignalRead

